src/Database/Expression/CommonTableExpression.php

```diff
@@ -0,0 +1,297 @@
+<?php
+declare(strict_types=1);
+
+namespace Cake\Database\Expression;
+
+use Cake\Database\ExpressionInterface;
+use Cake\Database\ValueBinder;
+use Cake\Error\Debugger;
+use Closure;
+use InvalidArgumentException;
+
+/**
+ * An expression that represents a common table expression definition.
+ */
+class CommonTableExpression implements ExpressionInterface
+{
+    /**
+     * The CTE name.
+     *
+     * @var string
+     */
+    protected $name;
+
+    /**
+     * The field names to use for the CTE.
+     *
+     * @var \Cake\Database\ExpressionInterface[]|string[]
+     */
+    protected $fields = [];
+
+    /**
+     * The modifiers to use for the CTE.
+     *
+     * @var \Cake\Database\ExpressionInterface[]|string[]
+     */
+    protected $modifiers = [];
+
+    /**
+     * The CTE query definition.
+     *
+     * @var \Cake\Database\ExpressionInterface|string
```

Comment:
What is the use case for supporting raw query strings?

Reply:
I didn't really had one in mind, I've just added it because of all the other places in the query builder and the expressions that accept raw SQL additionally to expressions.

Reply:
If we want to go a more stricter route and require expressions, then I'm all in :)

Reply:
Sounds like a 5.x feature to drop string support over objects for all expressions?

Reply:
SQL snippets are useful when getting the querybuilder to generate the SQL you want is to convoluted or too complicated. With that said we could/should use a wrapper object so that one can't accidentally enable raw-sql mode.

Reply:
Makes sense, and creating `QueryExpression` objects is easy enough.

Reply:
done